File: 알고리즘/Algorithm/Algorithm/quick.cpp
#include <iostream>
#include <vector>
using namespace std;

int arr[7] = { 38, 27, 43, 9, 3, 82, 10 };

void quick_sort(int start, int end) {
    if (start >= end) return;
    int pivot = start;
    int low = start + 1, high = end;

    while (low <= high) {
        while (arr[low] <= arr[pivot] && low <= end) low++;
        while (arr[high] >= arr[pivot] && high > start) high--;

        if (low > high)
            swap(arr[pivot], arr[high]);
        else
            swap(arr[low], arr[high]);
    }
 
    quick_sort(start, high - 1);
    quick_sort(high + 1, end);
}

void print() {
    for (const int& n : arr) cout << n << ' ';
}

int main() {
    quick_sort(0, 6);
    print();

    return 0;
}

File: Algorithm/Algorithm_JAVA/src/Quick.java
import java.util.Arrays;

public class Quick {
    public void quickSort(int[] data, int start, int end) {
        if (start >= end) return;
        int pivot = start;
        int low = start + 1, high = end;
        int temp;

        while(low <= high) {
            while(low <= high && data[low] < data[pivot]) low++;
            while(high > start && data[high] >= data[pivot]) high--;

            if(low > high) {
                temp = data[high];
                data[high] = data[pivot];
                data[pivot] = temp;
            } else {
                temp = data[low];
                data[low] = data[high];
                data[high] = temp;
            }
        }
        quickSort(data, start, high - 1);
        quickSort(data, high + 1, end);
    }

    public void print(int[] data) {
        for(int i=0; i < data.length; i++) {
            System.out.print(data[i] + " ");
        }
    }
}
